### ISSUE TITLE
Migrate to PDO for darabse access

### ISSUE THREAD

**Issue by mystralkk** [Geeklog-Core/geeklog#1092]
Currently, Geeklog uses mysqli and pgsql drivers through a wrapper layer.  However, you cannot use prepare statements or bind values to them.  You have to build an SQL statement manually like:
```PHP
$rows = [];
$sql = "SELECT * FROM some_table WHERE id ='" . DB_escapeString($id) . "'";
$result = DB_query($sql);

if (!DB_error() {
    while (($row = DB_fetchArray($result, false) !== false) {
        $rows[] = $row;
    }
}
```

This is tiresome and easily leads to errors.  With PDO as the backend, you could do like this:
```PHP
$rows = [];
$sql = "SELECT * FROM some_table WHERE id = :id";
$statement = DB_prepare($sql);
$statement->bindValue(':id', $id);

if ($statement->execute()) {
    foreach ($statement->fetchAll() as $row) {
        $rows[] = $row;
    }
}
```
How does this sound?

**Comment by eSilverStrike** [comment]
I am not that familiar with PDO. PDO is also faster is it not?

I assume both ways would be accessible to support plugins that have not been migrated yet?

**Comment by eSilverStrike** [comment]
Sounds good. 

As I said I am not to familiar with PDO. 

Is PDO automatically included in PHP installs or will we have to add some checks for it?

I guess the Geeklog install and Wiki Docs, etc. will also have to be updated

I really should find an article and read up on it.

If you think it makes sense to migrate Geeklog to it then I don't see me having a problem with it.

**Comment by mystralkk** [comment]
PDO is available by default( https://www.php.net/manual/en/pdo.installation.php ), but you have to enable a database-specific driver like pdo_mysql or pdo_pgsql.  This is easy to check in the installer.

**Comment by mystralkk** [comment]
After some research, I have found it is difficult to implement DB_numRows() with PDO, since there is no equivalent method.  I believe some of DB_numRows calls can be removed in some way, but there are more than 300 places where DB_numRows is called in the core and plugins shipped with the core ...

**Comment by eSilverStrike** [comment]
I guess rowcount for PDO is not an option since it doe not work with SELECT. I wonder why they do not have a solution out of the box for this. It seems like something most people would want when coding...

// A few possible solutions but most seem to deal with adding a COUNT to the select statement which doesn't really work in our situation...
https://stackoverflow.com/questions/54283485/pdo-row-count-num-rows-equivalent
https://stackoverflow.com/questions/11305230/alternative-for-mysql-num-rows-using-pdo

A lot of DB_numRows used in Geeklog are just checking to see if any records are returned before it does a fetch so I guess that could be a new function  called something like DB_rowsReturned for this use type.

Where Geeklog and plugins rely on numrows to actually cycle through the records I guess we would have to change it to a loop like:

`while ($row = DB_fetchArray($result)) {`

Of course all those changes are a lot of work and testing.

Do you thing all that work is worth it if an actual DB_numRows solution is not found???

How would plugins be dealt with that do not support PDO (yet)?

**Comment by mystralkk** [comment]
> Do you thing all that work is worth it if an actual DB_numRows solution is not found???

I don't think so.  The main reason I suggested introducing PDO is I would like to use prepared query and placeholders in SQL statements.  After I looked into mysqli_* and pg_* functions, I found these features are possible to implement with these functions.  In that case, I would create DB_prepare(string $sql), DB_bindValue(string $name, mixed $value), DB_execute($statement) functions without introducing the PDO extension.

> How would plugins be dealt with that do not support PDO (yet)?

If we decide to introduce PDO, I will make 2 APIs; one is the traditional DB_* functions (implemented with PDO) and a thin wrapper of PDO.

**Comment by hostellerie** [comment]
I think this issue is still relevant, but I would suggest slightly changing its goal.

The main benefit discussed here is not really PDO itself. The important improvement would be to add prepared statements and parameter binding to the Geeklog Database API.

A full migration from the current `mysqli` / PostgreSQL drivers to PDO would introduce a significant compatibility and testing burden, especially because Geeklog relies heavily on the existing `DB_*` API and on functions such as `DB_numRows()`.

The discussion already identified this problem, and the last comments also pointed out that prepared statements can be implemented directly with the native `mysqli_*` and `pg_*` APIs.

For this reason, I would suggest keeping the existing database abstraction layer and extending it with something like:

```php
DB_prepare()
DB_execute()
DB_bindValue()
```

or possibly an API where the values are passed directly to `DB_execute()`.

For example:

```php
$stmt = DB_prepare(
    "SELECT * FROM {$_TABLES['stories']} WHERE sid = ?"
);

$result = DB_execute($stmt, [$sid]);

while ($A = DB_fetchArray($result)) {
    // ...
}
```

The existing functions could remain fully supported:

```php
DB_query()
DB_fetchArray()
DB_numRows()
DB_escapeString()
```

This would preserve compatibility with existing plugins while allowing the Core and newer plugins to progressively adopt prepared statements.

I would see the migration happening in stages:

1. Add prepared statement support to the existing DB API.
2. Use it for all new Core code.
3. Gradually migrate security-sensitive and frequently modified queries.
4. Encourage plugin developers to use the new API.
5. Only reconsider the underlying driver implementation later if there is a clear benefit.

This would also improve security and code review. With manually constructed SQL, reviewers have to verify that every variable has been correctly escaped. With prepared statements, the separation between SQL and values becomes much clearer and reduces the risk of accidental SQL injection.

So instead of making PDO the primary objective, I would suggest reframing this issue as:

**“Add prepared statement support to the Geeklog Database API”**

PDO could still remain a possible internal implementation in the future, but it would no longer be a prerequisite.

This approach seems much less disruptive and provides most of the practical benefits originally expected from the PDO migration.
